stdlib/public/core/SetStorage.swift

```diff
@@ -127,6 +127,22 @@ internal class __EmptySetSingleton: __RawSetStorage {
 #endif
 }
 
+#if $Embedded
+public var _swiftEmptySetSingleton: (Int, Int, Int, Int, UInt8, UInt8, UInt16, UInt32, Int, Int, Int) =
```

Comment:
My understanding is that only homogeneous tuples are guaranteed to be layout-stable… Do we have a guarantee about how this will be laid out?

Reply:
it is paired to the same source that defines that - so ostensibly we could have a unit test that validates the expected layout and usage of such.

Reply:
Wait, why is the original C-based solution not suitable to use in embedded platforms?

Making this a Swift `var` is going to be risky even if we can somehow assume that tuple layout will match -- the empty singleton ought not be defined in writable memory. (It invariably leads to someone figuring out how to insert something into the empty singleton, leading to a really "fun" debugging session for somebody else. 😛)

Reply:
This is because in embedded Swift, the whole stdlib is only compiled to SIL and distributed as a .swiftmodule, and there isn't any .a or .o with actual code produced. Only the client's / app's build actually IRGen code, including the stdlib code. This lets us allow users to specify their actual LLVM ABI and IRGen flags without the need to have a matching stdlib build (this is a desirable property because embedded hardware/environments/SDKs have tons of actively used combinations of ABI settings). Essentially a similar outcome that you get with "header-only" C/C++ libraries.

Flip side is that we can't have C/C++ code in the embedded stdlib. The two concerns mentioned (layout guarantees, const-ness / non-writable memory) seem manageable to me, I'll try to address them separately afterwards (by adding tests checking the layout, and by exploring IRGen flags or other options that would make the singleton variables constant), if that's okay with you.

Added a comment explaining this in the meantime.

Reply:
I have another (though weaker) point to make: We should start considering moving parts of the *regular* Swift runtime from C to Swift, and the embedded Swift runtime can serve as a nice proving ground for that.

Reply:
Re. non-writable memory: AFAICT, the current regular definitions of the set/array singletons are *also* not in non-writable memory, and they are technically in writable memory. (Sounds like we should fix that there too.)